### ISSUE TITLE
Interface with in-engine heightmap generation tools - Non-destructive workflow

### ISSUE THREAD

**Issue by SlashScreen** [outobugi/Terrain3D#129]
I have a somewhat-functional tool for nondestructive terrain editing using stamps, similar to that of the Microverse unity plugin, currently interfacing (buggily) with HTerrain. It generates a heightmap as the process' output. Could there be ways to set regions / update an entire heightmap from GDScript?

**Comment by TokisanGames** [comment]
Yes. Look at importer.gd which is imports user files into the terrain via gdscript.

**Comment by SlashScreen** [comment]
you mean via `img = Terrain3DStorage.load_image(height_file_name, ResourceLoader.CACHE_MODE_IGNORE, r16_range, r16_size)` (line 62)? Also, the brush tool must only update a region, right? I'd like to be able to offer real-time feedback in the editor as the user drags the stamps around. How bad of an idea would it be to import an entire new heightmap as the user drags the stamp around, or would it only be possible if I were to calculate diffs somehow between the old heightmap and the new and only update that region? I know that's the Proper way to do it, but that's difficult.

**Comment by TokisanGames** [comment]
Look at the online help in godot (Shift+F1) for Terrain3DStorage or look at the functions exposed in _bind_methods here:
https://github.com/outobugi/Terrain3D/blob/d42c79a47db40db9c2dfce2eedf822ec3203b8df/src/terrain_3d_storage.cpp#L1563

`load_image` will import an image and place it on the terrain. You can import as small as one 1k region (though it currently also expects and overwrites control/color maps as well. We can change that to only import what is provided).

`get/set_map_region` will allow you to retreive and set a 1k region image at a time. If you're going to do it non-destructively you'll have to retain what was there before. Our sculpt tools work on whatever is in these images. It will be challenging to do it this way. 

Also you need to work across region boundaries.

It will probably be easier to incorporate a non-destructive data layer in C++ for stamps and curves by design.

**Comment by TokisanGames** [comment]
Currently on each edit update, we consolidate all heightmaps into one array to be sent to the shader.

Non-destructive editing could look like this:
* Heights only, generally
* A container for heightmap adjustments. It wouldn't need full coverage. It could be an array of images and their locations. 
* On each update_maps(), these get blit on to the generated heightmap and sent to the shader. 
* All image editing happens to the original layer as normal.
* Allows alpha stamps, which could be made permanent.
* Allows curves for roads and rivers #280. Curves can be defined that generate a road/riverbed stamp along the path, with an option for width, or single (riverbed) or dual grooves (road) 
* All stamps can have an feathered edge parameter
* A local node that can be worked on in a subscene. It allows for a sculpted terrain and texturing. In the main scene this node is treated like an alpha stamp. https://discord.com/channels/691957978680786944/1065519581013229578/1400475000070213715

**Comment by SlashScreen** [comment]
Data is sent to the GPU [here](https://github.com/outobugi/Terrain3D/blob/fd44f99a4bcbd4ab3106ebca3626d0387aa30364/src/terrain_3d_storage.cpp#L219C43-L219C43)?

**Comment by TokisanGames** [comment]
Yes, for the height/control/color maps. Actually sent when we set the material parameter.

**Comment by dev-bre** [comment]
Is this feature on the roadmap? is there any date when this would be available in the plugin?

**Comment by TokisanGames** [comment]
Whenever someone needs it and implements it. We don't need it for Out of the Ashes, so it's not on my personal task list anytime soon. There are far more pressing issues. I can provide guidance.

**Comment by PixelDough** [comment]
Microverse is one of my necessary addons when working in Unity. It's a really important feature to my workflow, so I'm interested in working on adding this functionality. I'm fairly new to Godot and don't know much about Godot's C++ plugin system, but if you have any advice and guidance on how to approach integrating this feature, I'd love to help out!
